Define types with more detail

nemchik opened the pull request:
- For things like `DecimalSeparator` and others I reviewed `jobrouter.js` and found the values that are accepted
- Avoid using `any`, instead define specific expected types
- Avoid using generic `Object`, instead define specific objects as much as possible
  - Some object properties have an `unknown` type (although if more context is available as to what is actually returned it would be ideal to define these more specifically)
- Don't define aliases for common types (ex: `type ElementId = string;`)
- Be consistent with parameter naming
  - `columnId` changed to `columnName` for consistency
  - `jr_show: (obj: string | string[]) => void;` changed to `jr_show: (id: string | string[]) => void;` because `obj` implies we should be passing in an object

A potential future option would be to add `dependencies` to `package.json` for `@types/jquery` and rely on and extend the underlying types as needed, which would be more accurate in cases where `$(...).val();` is returned from a `jr_...` function. This is more of a side-note for discussion.

nemchik committed:
Define types with more detail

jmerchel commented on api.d.ts:
```diff
@@ -1,80 +1,115 @@
 /********************************************************
- JobRouter JavaScript-API Stubs v5.1
+ JobRouter JavaScript-API Stubs v2022.1
```
Nice catch! We forgot to update this version number for quite some time. I think we will remove it in the future as it does not provide any real value.

jmerchel commented on api.d.ts:
```diff
@@ -1,80 +1,115 @@
  ********************************************************/
 type CustomErrorHandler = (errorMessage: string) => void;
-type ElementId = string;
+type DecimalSeparator = "" | "," | ".";
```
The DecimalSeparator and ThousandSeperator types are a great addition.

jmerchel commented on api.d.ts:
```diff
@@ -1,80 +1,115 @@
+    color: string;
+    onStamp?: (stamp: Stamp, x: number, y: number) => void;
 };
+type StampOptions = {
```
The Stamp and StampOptions type have a lot in common. Maybe we could use a combined type for Stamp to avoid the redundance. 
Like `type Stamp = { text: string } & StampOptions`

nemchik replied:
Done in https://github.com/jobrouter/javascript-api/pull/4/commits/253d2ce7fe9045f52d93ef3abead156924668c94

jmerchel commented on api.d.ts:
```diff
@@ -1,80 +1,115 @@
+};
+type ElementValue = string | number | Date | boolean;
+type MessageData = {
+    [key: string]: ElementValue;
```
The Naming of the type ElementValue suggests that it got something to do with a dialog element. This is a good fit for functions like jr_get_value but does not make much sense for replacements in a processmessage. 
We should either use a more generic name or dont use the ElementValue type here.

nemchik replied:
I had used `ElementValue` since it is the return type of `jr_get_value` which is used in the example in the documentation, but after reading more into it I think this is wrong.

The contents of jobrouter.js for this function are as follows: (if this snip of code needs to be redacted, I'm happy to remove it, just wanted to explain below)

```js
// Message - Functions
// ====================

var jr_message = function(message, data, customHandler) {
    var retValue = "";
    if (window.$JRMESSAGE[message]) {
        if (data) {
            retValue = window.$JRMESSAGE[message];
            $H(data).each(function(a) {
                $H(data).each(function(a) {
                    // Eckige Klammer notation ist nicht mehr offiziell dokumentiert,
                    // wird aber weiter unterstÃ¼tzt um alte Prozesse lauffÃ¤hig zu halten.
                    retValue = retValue.replace("[" + a.key + "]", a.value);
                    retValue = retValue.replace("#{" + a.key + "}", a.value);
                });
            });
        } else {
            retValue = window.$JRMESSAGE[message];
        }

        alert(jQuery("<div />").html(retValue).text());

    } else {
        throw new window.JobRouterException("No message '" + message + "' found!", customHandler);
    }
};
```

This implies to me that only `string` types should be accepted. The documentation for `.replace` from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace indicates it should be a string.

The JobRouter documentation reads:

> Optional: Object containing parameters to be substituted in the text.
> Example for filling place-keepers [var1] and [var2] in a process message:
> `{var1:"Wert1", var2:"Wert2"} or {var1:"Wert1", var2:jr_get_value("limit")}`

This shows using `jr_get_value("limit")` which can return a string, but can also return other types, which would need to be handled (converted to a string) in order to work as expected. It may make sense to update the JobRouter documentation for `jr_message` to read something like: `{var1:"Wert1", var2:"Wert2"} or {var1:"Wert1", var2:String(jr_get_value("limit"))}` but this doesn't cover things like formatting a date.

I've used `string` in https://github.com/jobrouter/javascript-api/pull/4/commits/253d2ce7fe9045f52d93ef3abead156924668c94

jmerchel commented on api.d.ts:
```diff
@@ -1,80 +1,115 @@
 declare const jr_subtable_refresh: (subtableViewName: string, columnName?: string, rowId?: number | "*", successCallback?: (subtableViewName: string, columnName: string, rowId: number | "*") => void, errorCallback?: (subtableViewName: string, columnName: string, rowId: number | "*", error: string) => void, customHandler?: CustomErrorHandler) => void;
-declare const jr_message: (message: string, data?: Object, customHandler?: CustomErrorHandler) => void;
-declare const jr_get_message: (message: string, data?: Object, customHandler?: CustomErrorHandler) => string;
+declare const jr_message: (message: string, data?: MessageData, customHandler?: CustomErrorHandler) => void;
```
I think data is a bad name for a parameter as it is really generic. Maybe we could rename this to something like replacements and also change the naming of your new type to Replacements?

nemchik replied:
Done in https://github.com/jobrouter/javascript-api/pull/4/commits/253d2ce7fe9045f52d93ef3abead156924668c94

jmerchel commented on api.d.ts:
```diff
@@ -1,80 +1,115 @@
+};
+type SubtableData = {
+    [key: number]: {
+        [key: string]: ElementValue;
```
We could also reuse your new SubtableRowObject type here to avoid duplication.

nemchik replied:
Done in https://github.com/jobrouter/javascript-api/pull/4/commits/253d2ce7fe9045f52d93ef3abead156924668c94

jmerchel reviewed (changes requested):
Thanks for this pull request. Please see my comments for some further improvements.
Regarding the jQuery dependency:
We are aiming to remove the jQuery dependency in the future therefore i think it would be better to not rely on their types.

nemchik committed:
Reduce redundancy

nemchik commented:
Two additional comments about https://github.com/jobrouter/javascript-api/pull/4/commits/253d2ce7fe9045f52d93ef3abead156924668c94

I added https://github.com/jobrouter/javascript-api/blob/253d2ce7fe9045f52d93ef3abead156924668c94/api.d.ts#L28 so that the function definition could be more easily used, allowing the function to be defined separately and then assigned to the value of the `onStamp` parameter.

I renamed `numberOfRows` to `rows` here https://github.com/jobrouter/javascript-api/blob/253d2ce7fe9045f52d93ef3abead156924668c94/api.d.ts#L73 since this parameter accepts a number, object, or array of objects.
